### PR TITLE
Prevent scheduler settings to be removed (master)

### PR DESCRIPTION
**Pull request by imobachgs**
Merge #17 into `master` branch:

* Fixes [bsc#968192](https://bugzilla.suse.com/show_bug.cgi?id=968192).
* SystemSettings refactoring.
* Added coveralls.

### CODE REVIEW

File: src/modules/SystemSettings.rb [hunk 1]
```diff
@@ -15,210 +15,296 @@ module Yast
...
+    KERNEL_SYSRQ_FILE = "/proc/sys/kernel/sysrq".freeze
+
+    # Return kernel configuration value for SysRq keys
+    #
+    # The value is read from /proc/sys/kernel/sysrq
+    #
+    # @return [String] Configuration value; returns an empty string if it's not set.
+    def kernel_sysrq
+      return @kernel_sysrq if @kernel_sysrq
+      content = SCR.Read(path(".target.string"), KERNEL_SYSRQ_FILE).to_s
```
Comment (lslezak) [inline src/modules/SystemSettings.rb]: NP: Maybe we could read the file directly using `File.read(KERNEL_SYSRQ_FILE)`, the SCR chroot feature does not make much sense if you want to read `/proc/` from kernel...

Comment (lslezak) [inline src/modules/SystemSettings.rb]: On the other hand this is more consistent with the rest, so I do not have a strong opinion here...

Comment (jreidinger) [inline src/modules/SystemSettings.rb]: if you want to read it and respect SCR chroot you can use https://github.com/yast/yast-yast2/blob/8eb9214265fdf8afcae1531c5a0309799028d79d/library/system/src/lib/yast2/target_file.rb
so `Yast::TargetFile.read(KERNEL_SYSRQ_FILE)`

File: src/modules/SystemSettings.rb [hunk 2]
```diff
@@ -15,210 +15,296 @@ module Yast
...
+    # @see enable_sysrq
+    def activate_sysrq
+      if enable_sysrq.nil? || enable_sysrq !~ /\A[0-9]\z/
+        log.warn("Not activating invalid ENABLE_SYSRQ value: #{enable_sysrq}")
+        return
       end
 
-      true
+      log.info("Activating SysRq config: #{enable_sysrq}")
+      SCR.Execute(path(".target.bash"), "echo '#{enable_sysrq}' > /proc/sys/kernel/sysrq")
```
Comment (lslezak) [inline src/modules/SystemSettings.rb]: NP: Maybe simple `File.write(KERNEL_SYSRQ_FILE, enable_sysrq)` :question: 
(I'm not sure if the new line is required, maybe `enable_sysrq + "\n"` is needed...)

File: src/modules/SystemSettings.rb [hunk 3]
```diff
@@ -15,210 +15,296 @@ module Yast
...
+      # set the scheduler
+      Bootloader.modify_kernel_params("elevator" => new_elevator)
+      # set bootloader configuration as 'changed' (bsc#968192)
+      Bootloader.proposed_cfg_changed = true
+
+      # activate the scheduler for all disk devices
+      return if new_elevator == :missing
+      Dir["/sys/block/*/queue/scheduler"].each do |f|
+        log.info "Activating scheduler '#{new_elevator}' for device #{f}"
+        File.write(f, new_elevator)
```
Comment (lslezak) [inline src/modules/SystemSettings.rb]: Ah, here you use `File.write` so the previous comment is then NP -> NTH.

File: src/modules/SystemSettings.rb [hunk 4]
```diff
@@ -15,210 +15,296 @@ module Yast
...
     end
 
-    # Kernel param 'elevator'
-    def GetIOScheduler
-      @elevator
+    # Activate IO scheduler
+    #
+    # @see activate_scheduler
+    def activate_scheduler
+      return if GetIOScheduler().nil?
```
Comment (lslezak) [inline src/modules/SystemSettings.rb]: NP: I guess simple `return unless GetIOScheduler()` should be enough as only `false` and `nil` evaluate to `false` (empty string evaluates to true).

File: src/modules/SystemSettings.rb [hunk 5]
```diff
@@ -15,210 +15,296 @@ module Yast
...
+      log.info("Activating scheduler: #{new_elevator}")
+      # set the scheduler
+      Bootloader.modify_kernel_params("elevator" => new_elevator)
+      # set bootloader configuration as 'changed' (bsc#968192)
+      Bootloader.proposed_cfg_changed = true
+
+      # activate the scheduler for all disk devices
+      return if new_elevator == :missing
+      Dir["/sys/block/*/queue/scheduler"].each do |f|
+        log.info "Activating scheduler '#{new_elevator}' for device #{f}"
```
Comment (lslezak) [inline src/modules/SystemSettings.rb]: NP: Missing parenthesis around the argument, make it consistent with the rest.

File: src/modules/SystemSettings.rb [hunk 6]
```diff
@@ -15,210 +15,296 @@ module Yast
...
 
-        @elevator = io_scheduler
-      else
-        Builtins.y2error(
-          "unknown IO scheduler '%1', use: %2",
-          io_scheduler,
-          GetPossibleElevatorValues()
+      # Set IO scheduler
+      SetIOScheduler(current_elevator)
+      Builtins.y2milestone("Global IO scheduler: %1", GetIOScheduler())
```
Comment (lslezak) [inline src/modules/SystemSettings.rb]: NP: Change to `log.info` when you are touching the code anyway...

File: src/modules/SystemSettings.rb [hunk 7]
```diff
@@ -15,210 +15,296 @@ module Yast
...
-    publish :function => :SetSysRqKeysEnabled, :type => "void (boolean)"
+    # Write IO Scheduler settings
+    #
+    # This method only has effect during normal mode. During installation,
+    # bootloader configuration is written at the end of the first stage.
+    #
+    # @see Bootloader#Write
+    # @see Write
+    def write_scheduler
+      return true unless Mode.normal
```
Comment (lslezak) [inline src/modules/SystemSettings.rb]: NP: Use simpler
```ruby
Bootloader.Write if Mode.normal
true
```

File: test/system_settings_test.rb [hunk 8]
```diff
@@ -0,0 +1,361 @@
+#!/usr/bin/env rspec
+
+require_relative "test_helper"
+
+Yast.import "SystemSettings"
+
+describe "Yast::SystemSettings" do
+  Yast.import "Bootloader"
```
Comment (lslezak) [inline test/system_settings_test.rb]: NP: Any reason to have the `import` here? Could you move it above so both `import` calls are together?

File: test/system_settings_test.rb [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+
+  subject(:settings) { Yast::SystemSettings }
+
+  before { settings.main }
+
+  describe "#GetPossibleElevatorValues" do
+    it "returns an array with possible schedulers" do
+      expect(settings.GetPossibleElevatorValues).to eq(["cfq", "noop", "deadline"])
```
Comment (lslezak) [inline test/system_settings_test.rb]: NP: Comparing Arrays is a bit tricky, if the order changes the test will fail. I usually do
```ruby
expect(settings.GetPossibleElevatorValues.sort).to eq(["cfq", "noop", "deadline"].sort)
```

Comment (imobachgs) [inline test/system_settings_test.rb]: Yes, you're right. And you can also use `match_array`

File: test/system_settings_test.rb [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+
+    before do
+      settings.SetSysRqKeysEnabled(sysrq_keys)
+      settings.SetIOScheduler(scheduler)
+      allow(Yast::Bootloader).to receive(:modify_kernel_params)
+      allow(Yast::Bootloader).to receive(:proposed_cfg_changed=)
+      allow(Dir).to receive(:[]).with(/scheduler/).and_return([disk])
+    end
+
+    context "given that SysRq keys status is unknown" do
```
Comment (lslezak) [inline test/system_settings_test.rb]: NP: I'd remove the "given that ..." beginning, the context is "SysRq keys status is unknown", IMHO it's redundant.

File: test/system_settings_test.rb (hunk 8 above)
Comment (imobachgs) [inline test/system_settings_test.rb]: Just a "mental" reason (I tend to put there imports I use in mocks). But it does not make sense :) So I'll move it.

File: test/system_settings_test.rb [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+    end
+  end
+
+  describe "#Write" do
+    let(:sysctl_sysrq) { "0" }
+    let(:mode) { "normal" }
+
+    before do
+      allow(Yast::Mode).to receive(:mode).and_return(mode)
+      allow(Yast::SCR).to receive(:Read).and_call_original
```
Comment (lslezak) [inline test/system_settings_test.rb]: NP: Move this mock below, the standalone line does not make sense, it only makes sense with the other `allow(Yast::SCR)` below. So group them together.

(My original comment was that this line is useless, then I noticed the other mock later.)

File: test/system_settings_test.rb [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+
+  describe "#SetIOScheduler" do
+    context "when scheduler is a known one" do
+      it "sets the scheduler to the given value" do
+        settings.SetIOScheduler("cfq")
+        expect(settings.GetIOScheduler).to eq("cfq")
+      end
+    end
+
+    context "when scheduler is an unknown one" do
```
Comment (lslezak) [inline test/system_settings_test.rb]: "when the scheduler is unknown" ?

File: test/test_helper.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+    add_filter "/spec/"
+  end
+
+  # for coverage we need to load all ruby files
+  src_location = File.expand_path("../../src", __FILE__)
+  Dir["#{src_location}/{modules,lib}/**/*.rb"].each { |f| require_relative f }
+
+  # use coveralls for on-line code coverage reporting at Travis CI
+  if ENV["TRAVIS"]
+    require "coveralls"
```
Comment (lslezak) [inline test/test_helper.rb]: Thanks for adding the code coverage support! :+1:

File: test/system_settings_test.rb (hunk 9 above)
Comment (lslezak) [inline test/system_settings_test.rb]: Yes, `match_array` it's better.

Comment (jreidinger) [inline test/system_settings_test.rb]: or https://www.relishapp.com/rspec/rspec-expectations/v/3-0/docs/built-in-matchers/contain-exactly-matcher
so contain-exactly is also possible

**Comment by lslezak** [comment]
LGTM, thanks!

**Pull request merged**